### DISCUSSION TITLE
请留下你对编码规范的建议!

### DISCUSSION THREAD

**Discussion by ZhangHanDong** [Rust-Coding-Guidelines/rust-coding-guidelines-zh#1] [Announcements]
你对编码规范有什么建议？ 

欢迎反馈！

**Comment by andylizi** [comment]
感谢各位在这个项目中付出的心血！几个不成熟的小建议，还请多多指教。

1. [`G.CNS.02 不要断言常量布尔类型`](https://github.com/Rust-Coding-Guidelines/rust-coding-guidelines-zh/blob/e7e9d3b/src/safe-guides/coding_practice/consts.md#gcns02-%E4%B8%8D%E8%A6%81%E6%96%AD%E8%A8%80%E5%B8%B8%E9%87%8F%E5%B8%83%E5%B0%94%E7%B1%BB%E5%9E%8B)
    这条建议来自于 clippy，但我感觉它与 clippy 的原意存在一定出入。对于 [`assertions_on_constants`](https://rust-lang.github.io/rust-clippy/master/#assertions_on_constants) 的描述和样例，我的理解是它在建议直接使用 `panic!()`，而不是用 `assert!(false)` 这种约等于 `if !false { panic!() }` 的无谓写法。这与编码规范里给出的例子截然不同：
    ```rust
    const MIN_OVERFLOW: usize = 8192;
    const MAX_START: usize = 2048;
    const MAX_END: usize = 2048;
    const MAX_PRINTED: usize = MAX_START + MAX_END;
    assert!(MAX_PRINTED < MIN_OVERFLOW);
    ```
    这段代码的本意是维护常量的一个 invariant，确保它在未来修改时不会被无意中破坏——其实就是 [`static_assertions`](https://docs.rs/static_assertions/) 的用途。

    因此就这个具体用例来说，我认为它断言常量、enforce invariant 的本意是好的，只是应该换成 `static_assertions` 确保这个断言是在编译时发生的，避免等到运行时再 panic。

    就这条规范来说，我建议将反例换成 clippy 的 `assert!(false)` 等。
2. [`P.VAR.02 不要在子作用域中使用变量遮蔽功能`](https://github.com/Rust-Coding-Guidelines/rust-coding-guidelines-zh/blob/e7e9d3be42/src/safe-guides/coding_practice/variables.md#pvar02-%E4%B8%8D%E8%A6%81%E5%9C%A8%E5%AD%90%E4%BD%9C%E7%94%A8%E5%9F%9F%E4%B8%AD%E4%BD%BF%E7%94%A8%E5%8F%98%E9%87%8F%E9%81%AE%E8%94%BD%E5%8A%9F%E8%83%BD)
    我感觉这个建议在大量的情景中都不适用，例如说
    ```rust
    let buf = read_something();
    if let Ok(buf) = String::from_utf8(buf) {
        println!("buf is {}", buf);
        return;
    }

    println!("buf is {:?}", &buf);
    // do more things with buf
    ```
    这应该算做“较小的作用域内遮蔽较大的作用域里的变量”？但类似的写法不仅不会难以阅读，反而十分常见。
    （当然，就这个特定的 `Result` 例子来说，更恰当的写法确实是换成 `match`。但在 if 后面的代码比 if 里面长很多的情况，提早返回会更清晰一些。）
3. [`P.VAR.06 利用变量遮蔽功能保证变量安全使用`](https://github.com/Rust-Coding-Guidelines/rust-coding-guidelines-zh/blob/e7e9d3be42/src/safe-guides/coding_practice/variables.md#pvar06---%E5%88%A9%E7%94%A8%E5%8F%98%E9%87%8F%E9%81%AE%E8%94%BD%E5%8A%9F%E8%83%BD%E4%BF%9D%E8%AF%81%E5%8F%98%E9%87%8F%E5%AE%89%E5%85%A8%E4%BD%BF%E7%94%A8)
    ```rust
    let data = {
         let mut data = get_vec();
        data.sort();
        data // 后面不在改动，但代码语义上没有表现出来，并且编译器也无法确保后面不会被改变
    };
    // Here `data` is immutable.
    ```
    可能是我理解错了，但为什么这里说“语义上没有表现出来”“编译器无法确保”呢？外部的 let 绑定是不变的，因此语义上没有 `mut` 当然就是不变的意思，编译器当然也会确保一个不变的绑定不会被改变。除了风格上的差别外，我感觉两个例子在语义上应该是等价的。

    从风格的角度，我在 rust repo 的 `./{library,compiler}/**/*.rs` 里面 grep 了一下。除去测试之类的代码，rust 整个项目中用 `let ([^\s]+) = \1;` 去除可变性的情况有四处，其中三处都是只能这样用——它们的两个声明之间还间隔了大量无关代码，不适合包装到表达式块里面。只有一处是可以 trivially 换成 `let $id = $id;` 的（Vec 先 sort 再 dedup）。

    而 `let $id = { let mut $id = ...; $id };` 这种模式，我用 `let ([^\s]+) = \{[\s\n]+let mut \1` 正则搜出了 6 处，略多于编码规范里推荐的写法，但不显著。
4. [`P.MEM.Lifetime.01 生命周期参数命名尽量简单`](https://github.com/Rust-Coding-Guidelines/rust-coding-guidelines-zh/blob/e7e9d3be42/src/safe-guides/coding_practice/memory/lifetime.md#pmemlifetime01-%E7%94%9F%E5%91%BD%E5%91%A8%E6%9C%9F%E5%8F%82%E6%95%B0%E5%91%BD%E5%90%8D%E5%B0%BD%E9%87%8F%E7%AE%80%E5%8D%95)
    建议将这一条限定在只有一个 lifetime 的情况。正如反例所示，当上下文里只存在一个 lifetime 的时候，用 `'strlife` 这种标记确实显得十分啰嗦。但在两个、甚至两个以上的情况还写 `'a` `'b` 的话，一复杂起来谁还记得 `'a` `'b` 对应的分别是啥呢。lifetime 本来已经是 Rust 中最难 reason about 的概念之一了，它的命名还是不要太过吝惜比较好。

    此外，在 rust repo 里搜 `<'[a-z]{2,}` 找到了 19443 个结果，排除少数误识别的 `'static` 也不少了。要把 rust 项目里面这两万处代码全都归为“不符合规范”未免不太妥当。
5. [`P.UNS.SafeAbstract.01 代码中要注意是否会因为恐慌（Panic）发生而导致内存安全问题`](https://github.com/Rust-Coding-Guidelines/rust-coding-guidelines-zh/blob/e7e9d3be42/src/safe-guides/coding_practice/unsafe_rust/safe_abstract.md#punssafeabstract01---%E4%BB%A3%E7%A0%81%E4%B8%AD%E8%A6%81%E6%B3%A8%E6%84%8F%E6%98%AF%E5%90%A6%E4%BC%9A%E5%9B%A0%E4%B8%BA%E6%81%90%E6%85%8Cpanic%E5%8F%91%E7%94%9F%E8%80%8C%E5%AF%BC%E8%87%B4%E5%86%85%E5%AD%98%E5%AE%89%E5%85%A8%E9%97%AE%E9%A2%98)
    > 恐慌一般在程序达到不可恢复的状态才用，当然在 Rust 中也可以对一些实现了 UnwindSafe trait 的类型捕获恐慌。

    panic 这个词在前文“错误处理”一节，以及末尾的术语表里是不翻译的，建议统一。

**Reply by ZhangHanDong** [reply]
感谢反馈！我再斟酌一下

**Reply by ZhangHanDong** [reply]
> 这应该算做“较小的作用域内遮蔽较大的作用域里的变量”？ 

关于第二条这句。是的。这条规则本意是让开发者避免因为这种写法而导致 Bug （本来不应该被重新绑定，而是直接修改的情况，看下面简单示例代码） 。可能描述上需要再改进一下。

```rust
fn main(){
    let mut a = 0;
    {
        // do something
        a = 42;
        
        // bug
        // let a = 42;
    }
    
    a; // use a again
}
```

**Reply by ZhangHanDong** [reply]
关于第一条 ，你说的有道理。只是这个示例 `assert!(false)` 太简单了，我想找一个更具有实践代表性的用例，现在这个示例的问题你说的确实有点道理，后面考虑一下更换更好的示例。或者你也可以贡献一下。

**Reply by ZhangHanDong** [reply]
关于第三条

> 为什么这里说“语义上没有表现出来”“编译器无法确保”呢？

因为它在定义之初是 可变变量啊。整个代码并没有表现出 「只是临时修改，然后就不会改变」的语义。

外部的 data 是外部的，只不过这里恰好起个同名。内外的data是两个不同的变量。

“编译器无法确保” 还是因为它是可变，编译器没法帮你保证它后面就不变啊。你跟【正例】对比理解吧。

**Reply by ZhangHanDong** [reply]
另外，规范只是一种建议，不是 铁律。 
而且这个规范是 《安全编码规范》，这里面列出的规则，不是为了教学，是为了警示作用，让开发者在写代码的时候，可以多一些考虑。

**Reply by ZhangHanDong** [reply]
关于第四条。

我觉得你可能没看仔细，描述里有写原因：【因为生命周期参数的目的是给编译器使用，用于防止函数中出现悬垂引用。】

主要是给编译器检查使用，不是给开发者去记忆的，应该尽量避免干扰到业务代码。

**Reply by ZhangHanDong** [reply]
关于第五条。 这个会统一的，这类问题可以直接发 pr 帮助修复。

**Reply by ZhangHanDong** [reply]
感谢你的认真反馈！ 希望后面多多交流。

**Reply by andylizi** [reply]
关于第三条

> 因为它在定义之初是 可变变量啊。外部的 data 是外部的，只不过这里恰好起个同名。内外的data是两个不同的变量。

其实如果这么说的话，`let data = data;` 前后的 data 同样也是两个不同的绑定啊……两种写法在我看来不存在根本上的差别。

虽然【正例】里面并没有显式写出大括号，但作用域在编译时仍然是存在的。正如 [Rustonomicon](https://doc.rust-lang.org/nomicon/lifetimes.html) 中所说：

> One particularly interesting piece of sugar is that each let statement implicitly introduces a scope.
> ```rust
> let x = 0;
> let y = &x;
> let z = &y;
> ```
> The borrow checker always tries to minimize the extent of a lifetime, so it will likely desugar to the following:
> ```rust
> // NOTE: `'a: {` and `&'b x` is not valid syntax!
> 'a: {
>     let x: i32 = 0;
>     'b: {
>         // lifetime used is 'b because that's good enough.
>         let y: &'b i32 = &'b x;
>         'c: {
>             // ditto on 'c
>             let z: &'c &'b i32 = &'c y;
>         }
>     }
> }
> ```

每一条 let 都有一个作用域。


> “编译器无法确保” 还是因为它是可变，编译器没法帮你保证它后面就不变啊

不知这里的 “后面” 是指的什么，我的理解是类似这样：

```rust
let data = {
     let mut data = get_vec();
    data.sort();
    data
};

data = vec![];  // 报错，绑定的值不能修改
data.push(1);   // 报错，不可变绑定不能获取 &mut
```

这不就是「编译器保证了它后面就不变」嘛。

如果你说的 “后面” 是指表达式块里面的话，

```rust
let data = {
     let mut data = get_vec();
    data.sort();
    data
    // 上一行 data 已经返回了，这之后不可能再写别的语句，更别提修改了
    // data.push(1);
};
```

---

关于第四条

> 主要是给编译器检查使用，不是给开发者去记忆的，应该尽量避免干扰到业务代码。

当只有一个 lifetime 时或者省略 lifetime 时确实是这样，但当环境中存在多个 lifetime 的时候，具体哪个引用属于哪个 lifetime 确实是需要开发者去理解的啊……

举个比较夸张的例子，比如 Rust 编译器里的这段代码：

```rust
/// 'cg = the duration of the constraint generation process itself.
struct ConstraintGeneration<'cg, 'cx, 'tcx> {
    infcx: &'cg InferCtxt<'cx, 'tcx>,
    all_facts: &'cg mut Option<AllFacts>,
    location_table: &'cg LocationTable,
    liveness_constraints: &'cg mut LivenessValues<RegionVid>,
    borrow_set: &'cg BorrowSet<'tcx>,
    body: &'cg Body<'tcx>,
}

pub(super) fn generate_constraints<'cx, 'tcx>(
    infcx: &InferCtxt<'cx, 'tcx>,
    liveness_constraints: &mut LivenessValues<RegionVid>,
    all_facts: &mut Option<AllFacts>,
    location_table: &LocationTable,
    body: &Body<'tcx>,
    borrow_set: &BorrowSet<'tcx>,
) {
    let mut cg = ConstraintGeneration {
        borrow_set,
        infcx,
        liveness_constraints,
        location_table,
        all_facts,
        body,
    };

    for (bb, data) in body.basic_blocks().iter_enumerated() {
        cg.visit_basic_block_data(bb, data);
    }
}
```

这里不光给 lifetime 起了名字，甚至还专门加了条注释……用 `'a` `'b` `'c` 这还怎么读？

> 另外，规范只是一种建议，不是 铁律。

确实不应该将编码规范死板地当作铁律，但……

> 标识为P为原则（Principle）。编号方式为P.Element.Number。
> 标识为G为规则（Guideline）。编号方式为G.Element.Number。

「原则」的语气还是挺重的，尤其是在与 G 对比的情况下。

**Reply by ZhangHanDong** [reply]
关于生命周期这个规则：生命周期参数命名尽量简单

我觉得你举的这个例子就是做到简单了呀。 这条规则又不是说必须用 a b c

**Reply by ZhangHanDong** [reply]
原则和规则，只是辅助编码的。 整体没有大碍起到提醒的作用就可以了，如果每句话都深抠的话，其实没法做到很严谨，也没法做到适合所有情况。

**Reply by ZhangHanDong** [reply]
第三条那描述可以再修正一下

**Reply by andylizi** [reply]
> 生命周期参数的命名应该尽量简单，用 'a/ 'b/ 'c 即可，**不要将其命名为带语义的名称**。

**Reply by ZhangHanDong** [reply]
这个描述可以改改

**Reply by ZhangHanDong** [reply]
第三条我又看了下，可能是我没有描述清楚。

反例的写法，外部的 data 确实是不可变，你觉得正例反例写法上没差别，也有道理。编译器也都可以保证后面的 不变性。

但我想表达的是，反例的写法在语义上其实没有表达清楚“ 先改变，然后不变”那种顺序语义。反例是在不变里面夹杂了一个可变。

**Reply by andylizi** [reply]
怎么说呢，感觉这是个人喜好和侧重不同的问题吧。如果场景里面强调的是「先临时改变，然后不变」，那么正例的写法显然是最贴合这种语义的，在此类场景中也确实最为恰当：

```rust
let mut options = parse_args();

// 校验用户传入的参数，填充默认值，etc
// 此处省略100行代码

let options = options;

// 开始业务流程
```

但我觉得规范里举出的场景，强调的不是「先临时改变，然后不变」，而是「初始化 `data` 变量」。前者只是实现后者的途径之一，而不是问题的本质。就前者而言，上面那段代码才是比较符合的。

用 [Rust Reference](https://doc.rust-lang.org/stable/reference/statements.html#let-statements) 的话来说，`let` 语句等号后面的是一个 initializer expression，编译器一视同仁，因此下面两段代码的语义是一致的：

```rust
// 声明一个 `data` 变量，将其初始化为 `load_data()` 表达式的值
let data = load_data();
```

```rust
// 声明一个 `data` 变量，将其初始化为 `{ ... }` 表达式的值
let data = {
    let mut data = load_data();
    data.sort();
    data
};
```

这种写法有个优势，它能让读者一眼看明白，这个表达式块的目的就是为了产生（produce）data 的值，里面的状态之后就不再需要用到了。例如：

```rust
let names = {
    let mut names = parse_names();
    let families = load_families();
    names.sort_by_key(|n| families.get(n.last_name).map(|f| f.size()).unwrap_or(1));
    names
};
let data = {
    let mut data = load_data();
    let records = get_records();
    data.sort_by_key(|entry| records.get(entry).unwrap_or(&0));
    data
};
```

清晰地表达了「块里的临时变量就是用来初始化 `names` / `data` 的」。而如果全都写在一个作用域里面：

```rust
let mut names = parse_names();
let families = load_families();
names.sort_by_key(|n| families.get(n.last_name).map(|f| f.size()).unwrap_or(1));
let names = names;

let mut data = load_data();
let records = get_records();
data.sort_by_key(|entry| records.get(entry).unwrap_or(&0));
let data = data;

// 此处省略100行

// 等等，records 是啥？
```

（此外，一部分开发者没有分隔不同逻辑段落的好习惯。上面的代码，如果两段之间没有那个空行，那就更加没法看了）

---

反例的写法还有一个好处是，如果你不小心给 `data` 加了 `mut`，`let mut data = { ... };`，编译器会警告 variable does not need to be mutable。而正例的写法，

```rust
let mut data = get_vec();
// do something mutably with data

// 100 lines later

// do something immutably with data
```

编译器不会提示你，在第二行以后 `data` 就不需要可变了。换句话说，`let data = data;` 是「不必须的」——它既不影响代码的正确性，也不太影响代码的可读性（因为 mutable borrow 必须 exclusive 是编译器保证的），并且编译器和 clippy 都不会主动提醒你，需要你主动意识到「这个地方逻辑上应该加上 `let data = data;`」。况且这种写法，我相信对很多 Rust 不是第一语言的开发者（真的有人会用 Rust 入门吗……）来说是比较反直觉的。毕竟除了 Rust 外好像没有哪个常用语言可以这样写（？）。

---

总结一下的话，两个写法我觉得各有各的适用情形，各人有各人的喜好。就「初始化变量」的特定情形来说，前面论证了反例的写法有时还更有优势一些。如果我提 PR 的时候有人 request change 说 “建议这里改成 let something = something;“，那我的反应肯定是无语的。

**Reply by ZhangHanDong** [reply]
你说的第一条那个示例，应该是放到“例外”里，而不是反例，我忘记是不是因为我写错了。总之，已经修改，感谢 review。
